### PR TITLE
Update react config

### PR DESCRIPTION
**Pull request by edvinerikson**
Check the first commit

### CODE REVIEW

File: .react-eslintrc [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "react/jsx-indent": [ 2, 2 ],
     "react/jsx-key": 2,
     "react/jsx-max-props-per-line": [ 2, { "maximum": 4 }],
-    "react/jsx-no-bind": 2,
+    "react/jsx-no-bind": 0,
```
Comment (EnzoMartin) [inline .react-eslintrc]: Why? Last I read there were actual performance impacts from using bind

Comment (edvinerikson) [inline .react-eslintrc]: intention is to allow arrow functions not `.bind()`

Comment (edvinerikson) [inline .react-eslintrc]: hmm, maybe there is a option for only that

Comment (beckend) [inline .react-eslintrc]: Arrow function will still cause a re-render right

Comment (viart) [inline .react-eslintrc]: >intention is to allow arrow functions not .bind()
  
If I understand it properly, arrows functions ideologically should not be manually bound (they are "pre-bound" to the lexical scope for you already) ;)

File: .react-eslintrc [hunk 2]
```diff
@@ -41,16 +41,16 @@
...
     "react/no-direct-mutation-state": 2,
     "react/no-is-mounted": 2,
     "react/no-multi-comp": 2,
-    "react/no-set-state": 2,
+    "react/no-set-state": 0,
     "react/no-string-refs": 2,
     "react/no-unknown-property": 2,
     "react/prefer-es6-class": 2,
-    "react/prefer-stateless-function": 2,
+    "react/prefer-stateless-function": 0,
```
Comment (viart) [inline .react-eslintrc]: Why so?

File: .react-eslintrc [hunk 3]
```diff
@@ -61,7 +61,7 @@
       order: [
         "static-methods",
         "lifecycle",
-        "/^on.+$/",
```
Comment (viart) [inline .react-eslintrc]: Why?

File: .react-eslintrc [hunk 4]
```diff
@@ -41,16 +41,16 @@
...
     "react/no-deprecated": 2,
-    "react/no-did-mount-set-state": 2,
-    "react/no-did-update-set-state": 2,
+    "react/no-did-mount-set-state": 0,
+    "react/no-did-update-set-state": 0,
     "react/no-direct-mutation-state": 2,
     "react/no-is-mounted": 2,
     "react/no-multi-comp": 2,
-    "react/no-set-state": 2,
+    "react/no-set-state": 0,
```
Comment (viart) [inline .react-eslintrc]: Due to the fact that we use Redux, the local state should be rather an exception than a rule, therefore I would keep it as it was.

File: .react-eslintrc (hunk 1 above)
Comment (edvinerikson) [inline .react-eslintrc]: arrow functions are bound yes. but we still disallow them in jsx. `<div onCLick={() => this.setState({ clicked: true })}/>`

File: .react-eslintrc (hunk 3 above)
Comment (edvinerikson) [inline .react-eslintrc]: they are called `handle`

File: .react-eslintrc (hunk 4 above)
Comment (edvinerikson) [inline .react-eslintrc]: read the commit message, in case you haven't

File: .react-eslintrc [hunk 5]
```diff
@@ -41,16 +41,16 @@
     "react/jsx-uses-vars": 2,
     "react/no-danger": 2,
     "react/no-deprecated": 2,
-    "react/no-did-mount-set-state": 2,
-    "react/no-did-update-set-state": 2,
+    "react/no-did-mount-set-state": 0,
+    "react/no-did-update-set-state": 0,
```
Comment (viart) [inline .react-eslintrc]: Is it safe? I thought it was not recommended...

File: .react-eslintrc [hunk 6]
```diff
@@ -14,7 +14,7 @@
     "valid-jsdoc": 0,
     "jsx-quotes": [2, "prefer-double"],
     "react/display-name": [ 2, { "ignoreTranspilerName": false } ],
-    "react/forbid-prop-types": 2,
+    "react/forbid-prop-types": 0,
```
Comment (viart) [inline .react-eslintrc]: Why so?

File: .react-eslintrc (hunk 2 above)
Comment (edvinerikson) [inline .react-eslintrc]: they are not that good. I want to be able to use class components instead without disabling a rule in every file.

Comment (EnzoMartin) [inline .react-eslintrc]: Agreed with this. I've always thought this rule was stupid.

File: .react-eslintrc (hunk 5 above)
Comment (EnzoMartin) [inline .react-eslintrc]: Did mount should be safe, but did update seems really sketchy as that can result in an infinite loop..

Comment (edvinerikson) [inline .react-eslintrc]: yeah it's safe. Yeah this eslint rule does not recommends it but nothing else really.

Comment (beckend) [inline .react-eslintrc]: You can escape that loop if you mutate the state object directly on did update

File: .react-eslintrc (hunk 1 above)
Comment (viart) [inline .react-eslintrc]: Yes, since in your example the new unnamed function will be created each time the block is evaluated.

File: .react-eslintrc (hunk 6 above)
Comment (EnzoMartin) [inline .react-eslintrc]: I remember a conversation about how I thought it was stupid that we enforce specific prop specifications (though more so in regards to React handling it instead of a model) and @edvinerikson you were pretty dead set on it being needed, why the change of heart?

File: .react-eslintrc (hunk 1 above)
Comment (EnzoMartin) [inline .react-eslintrc]: I'd rather use a rule specifically to allow arrow functions than allow bind.

Even still, you can declare the function outside of the JSX block and pass it as a variable

Comment (edvinerikson) [inline .react-eslintrc]: @beckend 
> Arrow function will still cause a re-render right
  
no, not if you're smart where you use them or how the component is using them that you pass it to. they are not `===` equal but by default React would re-render anyway and since it's a function it's fine because the component using it would call `this.props.foo()` directly and not `this.memoFoo()`

File: .react-eslintrc (hunk 5 above)
Comment (EnzoMartin) [inline .react-eslintrc]: Ehhhhhh.. I think we should keep did-update to not allowed, but I'm fine with `did-mount` being allowed

File: .react-eslintrc (hunk 6 above)
Comment (viart) [inline .react-eslintrc]: I would rather keep it strict and rely on it until we decide and move to either Flow or TS.

File: .react-eslintrc (hunk 1 above)
Comment (EnzoMartin) [inline .react-eslintrc]: So basically, yes, they will, unless you know what you're doing.

File: .react-eslintrc (hunk 5 above)
Comment (edvinerikson) [inline .react-eslintrc]: you can't mutate the state directly in react.

from docs:
> componentDidUpdate() is invoked immediately after updating occurs. This method is not called for the initial render.

> Use this as an opportunity to operate on the DOM when the component has been updated. This is also a good place to do network requests as long as you compare the current props to previous props (e.g. a network request may not be necessary if the props have not changed).

File: .react-eslintrc (hunk 6 above)
Comment (edvinerikson) [inline .react-eslintrc]: No I've always thought it was bad not be able to use `React.PropTypes.object`.
Probably a miss-understanding then

File: .react-eslintrc (hunk 1 above)
Comment (viart) [inline .react-eslintrc]: >Even still, you can declare the function outside of the JSX block and pass it as a variable

Yep, that is a proper way. Either using `redux`'s `mapDispatch...` or "statefull" component.

Comment (edvinerikson) [inline .react-eslintrc]: it will be a bottleneck if you use shallow-equal checks for all props. don't do that for callbacks and you're fine.

File: .react-eslintrc (hunk 5 above)
Comment (viart) [inline .react-eslintrc]: >if you mutate the state object directly

and then after few months GL&HF with debugging of such code :D

Comment (EnzoMartin) [inline .react-eslintrc]: Yes, that says mutate the DOM, not set the state, which will cause a re-render and for did-update to trigger again

File: .react-eslintrc (hunk 4 above)
Comment (viart) [inline .react-eslintrc]: Which one?

File: .react-eslintrc (hunk 1 above)
Comment (EnzoMartin) [inline .react-eslintrc]: Yea I think we should keep this as an error, though I'm ok with allowing the use of arrow functions if possible

Comment (beckend) [inline .react-eslintrc]: You know redux-form is a example of the arrow thing is problematic, you have any random third party react-input-toggle-whatever-form component that wants a `onChange` function and you kinda need to use a arrow function to pass the playerId to the onChange and trigger something.

It would simplify a lot of onChange stuff but yea will probably will re-render.

File: .react-eslintrc (hunk 5 above)
Comment (edvinerikson) [inline .react-eslintrc]: fine with keeping did-update as error

File: .react-eslintrc (hunk 3 above)
Comment (viart) [inline .react-eslintrc]: Not always: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-handler-names.md

File: .react-eslintrc (hunk 6 above)
Comment (edvinerikson) [inline .react-eslintrc]: the issue is that it doesn't provide any value since you have to do it for all gql data props. you can, but it takes a lot of time creating that shape and it's not that useful in the end anyway.

Comment (EnzoMartin) [inline .react-eslintrc]: I would *love* to get rid of this, but it does serve a purpose as a sort-of model enforcement which we don't have otherwise, so, until we have something to replace that, we need to keep it

File: .react-eslintrc (hunk 2 above)
Comment (beckend) [inline .react-eslintrc]: They say stateless are better in performance, is it really true?

**Comment by edvinerikson** [comment]
https://github.com/LeoGears/codestyle/pull/52/commits/b30fad1bcf640f16d1ed668be7c7a8f0b48df860

File: .react-eslintrc (hunk 6 above)
Comment (beckend) [inline .react-eslintrc]: I want this strict, I'm tired of guessing what data is fed into components - i'm looking at you backbone.

File: .react-eslintrc (hunk 3 above)
Comment (beckend) [inline .react-eslintrc]: How about a regex which accepts both `onSomething` and `handleSomething`, they are used a lot in third party components.

File: .react-eslintrc (hunk 1 above)
Comment (edvinerikson) [inline .react-eslintrc]: if the child component uses props correctly it will not really be a problem.

```
handleCb = () => {
  this.props.onFoo();
}
```

this will always have a ref to latest props even if the component returned false in componentSHouldUpdate

File: .react-eslintrc (hunk 3 above)
Comment (EnzoMartin) [inline .react-eslintrc]: Why not both? 😛

File: .react-eslintrc (hunk 6 above)
Comment (viart) [inline .react-eslintrc]: So what is the point in them at all and/or moving to Flow/TS (those will enforce much stricter rules) if one doesn't want to document the format in detailed way? ;)

File: .react-eslintrc (hunk 2 above)
Comment (edvinerikson) [inline .react-eslintrc]: nope and they never said it have better perf either. they said stateless components will be able to unlock more perf opt paths. doesn't actually mean they will be faster.

File: .react-eslintrc (hunk 6 above)
Comment (EnzoMartin) [inline .react-eslintrc]: Well, Backbone does have this actually, the `parse` and `validate` methods can be implemented which it runs through every time there's a change, it's just that we don't implement it

File: .react-eslintrc (hunk 3 above)
Comment (edvinerikson) [inline .react-eslintrc]: `<Foo onFoo={this.handleFoo} />`

Comment (edvinerikson) [inline .react-eslintrc]: more common than `<Foo onFoo={this.onFoo} />`. been a "best-practice" since a year+ or something

File: .react-eslintrc (hunk 6 above)
Comment (beckend) [inline .react-eslintrc]: So lets start enforcing it please

Comment (edvinerikson) [inline .react-eslintrc]: PropTypes is legacy really. Flow have replaced them but they do still exist. however they don't accept feature requests for them anymore.

they are useful when using data that is not from gql but 99% cases it is data from gql.

Comment (edvinerikson) [inline .react-eslintrc]: should still use them for other types of data until Flow/TS. `PropTypes.string.isRequired` is fine. but using `PropTypes.shape()` to describe your data requirements twice is not efficient since 
```
gql`
  fragment foo on B {
    id
  }
`
```
does it pretty good

File: .react-eslintrc (hunk 5 above)
Comment (viart) [inline .react-eslintrc]: Why would you set state in `did-mount`? isn't it eq to initialState?

File: .react-eslintrc (hunk 6 above)
Comment (beckend) [inline .react-eslintrc]: If you have a way to soft-enforce the data - then fine, hoping people don't do
```js
type badType = {
  prop1: any,
  prop2: {
   prop3: any
  }
};
```

File: .react-eslintrc (hunk 3 above)
Comment (viart) [inline .react-eslintrc]: Agree.

Comment (edvinerikson) [inline .react-eslintrc]: @beckend 
3rd-partys can still use them, I am not talking about the propname I am talking about the name if the callback function. e.g `this.handleFoo` and not `<div onFoo={} />`

@EnzoMartin I don't think have used `this.on` anywhere

File: .react-eslintrc (hunk 5 above)
Comment (edvinerikson) [inline .react-eslintrc]: it's called after render. useful when needing state on client but not on server.

File: .react-eslintrc (hunk 2 above)
Comment (viart) [inline .react-eslintrc]: >prefer-stateless-function

Positive side effect of it that it's forcing you to think in redux-way (do not create local state...), which you should do in major cases.

File: .react-eslintrc (hunk 6 above)
Comment (viart) [inline .react-eslintrc]: What about our custom `PropTypes.GQL`? :)

**Comment by edvinerikson** [comment]
I feel that a lot of the react rules actually remove good things about React.

File: .react-eslintrc (hunk 6 above)
Comment (edvinerikson) [inline .react-eslintrc]: graphql-anywhere have that I think. not sure it works for everything. and still, what exact value does it provide to define twice? should still use shape for other objects that is not from gql but it doesn't mean `PropTypes.object` is bad in anyway.

File: .react-eslintrc (hunk 2 above)
Comment (edvinerikson) [inline .react-eslintrc]: > Positive side effect of it that it's forcing you to think in redux-way (do not create local state...), which you should do in major cases.

local state is much better than redux state in 99% of cases. read the commit message I linked.

File: .react-eslintrc (hunk 6 above)
Comment (viart) [inline .react-eslintrc]: It's not bad it's just ambiguous.

File: .react-eslintrc (hunk 2 above)
Comment (viart) [inline .react-eslintrc]: Since we decided to use Redux as a single point of truth we should stick to it. If you believe it doesn't work well for us, let's take it offline and either drop it or continue with it. Redux + local state == mass (unnecessary re-renders) in most cases.

Comment (edvinerikson) [inline .react-eslintrc]: Redux + local state is good. they serve different purposes

Comment (edvinerikson) [inline .react-eslintrc]: right now it A or B not A and B which is bad

Comment (viart) [inline .react-eslintrc]: >read the commit message I linked.

Would you be so kind to either add it to the description of PR of quote it here, since I don't know commit you're referring to? Thanks.

Comment (edvinerikson) [inline .react-eslintrc]: https://github.com/LeoGears/codestyle/pull/52#issuecomment-273432057

File: .react-eslintrc [hunk 7]
```diff
@@ -23,7 +23,7 @@
     "react/jsx-indent": [ 2, 2 ],
     "react/jsx-key": 2,
     "react/jsx-max-props-per-line": [ 2, { "maximum": 4 }],
-    "react/jsx-no-bind": 2,
+    "react/jsx-no-bind": [2, { "allowArrowFunctions": true }],
```
Comment (viart) [inline .react-eslintrc]: I'm still against of changing it, reasons are in the thread: https://github.com/LeoGears/codestyle/pull/52#discussion_r96593728

Comment (edvinerikson) [inline .react-eslintrc]: it's being used in the React tutorial on the react docs page.
https://facebook.github.io/react/tutorial/tutorial.html#an-interactive-component

they wouldn't do that if it was really bad for performance.

File: .react-eslintrc (hunk 4 above)
Comment (viart) [inline .react-eslintrc]: >Adds ability to use React state API.
>The intention with using Redux in an application is to eliminate consistency bugs and make it easier to share state.
>I have found that when using Apollo + gql there is not much state that needs to be shared between components and in this case Redux is actually making the application grow significantly (both logic and size) than compared to just use React state api directly.

I disagree. Auth, i18n, etc all they are shared.

File: .react-eslintrc (hunk 5 above)
Comment (viart) [inline .react-eslintrc]: Sounds like a very specific case to me aka "exception from the rules".

File: .react-eslintrc (hunk 4 above)
Comment (edvinerikson) [inline .react-eslintrc]: yes? they should be in redux then. not against that. but all state shouldn't be there. especially not state that is local to a component. React state API is still not bad, it's as good as redux is which is why both should be allowed

File: .react-eslintrc (hunk 5 above)
Comment (edvinerikson) [inline .react-eslintrc]: react state api should still be allowed without having to escape rules all the time. nothing bad with react state so it should be allowed.

File: .react-eslintrc (hunk 7 above)
Comment (viart) [inline .react-eslintrc]: So, even considering the fact they made a mistake (just to shorten an example) on showing that way and with understanding that each re-render *will* create 1+ unnecessary function we're gonna use it too, just because "facebook is doing so", am I understanding correct what you are trying to say?

File: .react-eslintrc (hunk 4 above)
Comment (viart) [inline .react-eslintrc]: IMO if you have such a special case that you need a local state (e.g. dropdown component with `isOpen) then it's an acceptable exception, just silent the linter there in all other cases - use Redux.

File: .react-eslintrc (hunk 7 above)
Comment (edvinerikson) [inline .react-eslintrc]: no... I am saying that re-creating a function is not affecting performance and that it will probably be the last thing we have to worry about when it comes to perf.

File: .react-eslintrc (hunk 2 above)
Comment (viart) [inline .react-eslintrc]: >right now it A or B not A and B which is bad

In terms of date there must just a single point of truth, local state if for VERY specific cases.

File: .react-eslintrc (hunk 4 above)
Comment (edvinerikson) [inline .react-eslintrc]: I don't agree. right now redux is the good citizen and react state api is bad and discouraged but actually they should actually be equal.

File: .react-eslintrc (hunk 2 above)
Comment (edvinerikson) [inline .react-eslintrc]: I would say Redux is for VERY specific cases as well.

**Comment by viart** [comment]
@edvinerikson seems like we can't reach an agreement of few points here, let's wait comments from @m64253 ;)

**Comment by EnzoMartin** [comment]
@viart @edvinerikson both of you should be more than capable of coming to an agreement together by compromising on certain points

File: .react-eslintrc (hunk 7 above)
Comment (EnzoMartin) [inline .react-eslintrc]: I'm good with this change allowing the arrow functions 👍

File: .react-eslintrc (hunk 6 above)
Comment (EnzoMartin) [inline .react-eslintrc]: We should let this be for now, let's revisit when we're ready to move to Flow/TS

**Comment by viart** [comment]
The things which I agree on I've highlighted already other ones in my opinion are opening rabbit holes, therefore I'm not ready to compromise on them.

File: .react-eslintrc (hunk 7 above)
Comment (viart) [inline .react-eslintrc]: [fresh one](https://ryanfunduk.com/articles/never-bind-in-render/)

File: .react-eslintrc (hunk 4 above)
Comment (EnzoMartin) [inline .react-eslintrc]: I've never liked our blanket ban on `setState`, I think it's better handled through PRs.

I agree that this makes it a bit hard to choose though, as there's good arguments for both cases.

I think in this case it's better to leave it as not allowed and continue using overrides where it makes sense to shortcircuit Redux

File: .react-eslintrc (hunk 7 above)
Comment (edvinerikson) [inline .react-eslintrc]: > performance because those are usually the components that are used hundreds of times in your interface such as table rows and the like.

File: .react-eslintrc (hunk 3 above)
Comment (EnzoMartin) [inline .react-eslintrc]: @edvinerikson Fair enough

File: .react-eslintrc (hunk 2 above)
Comment (EnzoMartin) [inline .react-eslintrc]: Redux aside, this rule is pretty annoying if you prefer using classes. We should disable it.

File: .react-eslintrc (hunk 7 above)
Comment (viart) [inline .react-eslintrc]: once we remove this rule the arrows will be everywhere ==
>performance because those are usually the components that are used hundreds of times in your interface such as table rows and the like.

One should use it in render once when it's absolutely necessary and he/she understands completely potential side-effects.

Comment (EnzoMartin) [inline .react-eslintrc]: Oh it applies to arrow functions too.. well, in that case, this rule should remain as an error, and we should continue to set arrow functions as variables

File: .react-eslintrc (hunk 2 above)
Comment (viart) [inline .react-eslintrc]: Sure, agree on this one too, since it's a way to handle "no bind in render".

File: .react-eslintrc [hunk 8]
```diff
@@ -41,16 +41,16 @@
     "react/jsx-uses-vars": 2,
     "react/no-danger": 2,
     "react/no-deprecated": 2,
-    "react/no-did-mount-set-state": 2,
+    "react/no-did-mount-set-state": 0,
```
Comment (viart) [inline .react-eslintrc]: Do we have a real case where we use `setState` in `didMount` atm?

File: .react-eslintrc (hunk 7 above)
Comment (edvinerikson) [inline .react-eslintrc]: arrow functions doesn't use `.bind()`

File: .react-eslintrc (hunk 8 above)
Comment (edvinerikson) [inline .react-eslintrc]: yes, already mentioned that in some thread. when you want state on client but not on server

File: .react-eslintrc (hunk 7 above)
Comment (EnzoMartin) [inline .react-eslintrc]: They have the same effect though when part of the JSX markup

File: .react-eslintrc (hunk 8 above)
Comment (EnzoMartin) [inline .react-eslintrc]: This is a safe rule to disable, as `did-mount` can only ever fire 1 time in the entire lifecycle of a component

File: .react-eslintrc (hunk 7 above)
Comment (edvinerikson) [inline .react-eslintrc]: they *may* have perf impact if used in lists. they will have DX impact if you have to split your render method into multiple smaller ones (eliminates one nice thing with React).

```jsx
<Match>
  {(params) => (
    <h1>Hello, {params.username}</h1>;
  )}
</Match>
```
vs 
`<Match>{this.renderWelcome}</Match>`


the later removes declarative aspect of React with imperative instead.

Comment (edvinerikson) [inline .react-eslintrc]: https://github.com/gutro/fulcrum-frontend/blob/321b2a7dc35635fc4aeddb75aac2da4e126242e9/src/web/desktop/Game/components/GamePage/index.js#L113-L204

Comment (viart) [inline .react-eslintrc]: I like the last one, and I believe in it `this.renderWelcome` should be `<Welcome />`, much declarative more fun :D

Comment (viart) [inline .react-eslintrc]: https://github.com/LeoGears/codestyle/pull/52#discussion_r96630559 OMG complexity of that code is over9000 (it is not readable) and must be split into sub-components at least.

Comment (edvinerikson) [inline .react-eslintrc]: > I like the last one, and I believe in it this.renderWelcome should be <Welcome />, much declarative more fun :D
  

so `(params) => <Welcome username={params.username} />`?

> #52 (comment) OMG complexity of that code is over9000 (it is not readable) and must be split into sub-components at least.


game launching is pretty complex. maybe add one abstraction. not sure

Comment (edvinerikson) [inline .react-eslintrc]: https://github.com/gutro/frontend/blob/master/apps/desktop/src/views/view.game-play.js#L258-L309

Comment (edvinerikson) [inline .react-eslintrc]: btw
https://blog.daftcode.pl/hype-driven-development-3469fc2e9b22#.oma8mlld9
> Step 3: Mania starts
All shades of hype driven developers read blog posts and attend conferences. Soon the teams all over the world start using the new technology. Due to the blurred message — some of them make hasty decision to use framework even though it does not solve any of their actual problems. Yet the team does have expectation that this new technology will help.


====== https://ryanfunduk.com/articles/never-bind-in-render/

Comment (viart) [inline .react-eslintrc]: Just compare the implementation to react-native's one (yes, they're bit different but not completely), the later doesn't have such "nesting hell".

Comment (edvinerikson) [inline .react-eslintrc]: mobile is quite simpler as well
https://github.com/gutro/fulcrum-frontend/blob/f17d15ca47eb246d9d37e2f11c9792aefa273fc9/src/web/mobile/Game/components/GamePage/index.js#L123-L186

**Comment by EnzoMartin** [comment]
Ok so it seems the only thing left here is:

- [] Leave no-bind as error

**Comment by edvinerikson** [comment]
updated

File: .react-eslintrc (hunk 4 above)
Comment (viart) [inline .react-eslintrc]: >I think in this case it's better to leave it as not allowed and continue using overrides where it makes sense to shortcircuit Redux

+1
